plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/EmitterMode.hs

```diff
@@ -68,3 +70,35 @@ logWithBudgetEmitter = EmitterMode $ \getBudget -> do
         let withBudget = logs <&> \str -> encodeRecord (str, exCpu, exMemory)
         modifySTRef logsRef (`DList.append` withBudget)
   pure $ CekEmitterInfo emitter (DList.toList <$> readSTRef logsRef)
+
+{- | Emits log and, when script evaluation fails, call trace.
+
+This requires script to be compiled with `PlutusTx.Plugin:profile-all` turned on because this relies
+on compiler-generated trace calls that notifies entrance and exit of a function call. These traces
+that mark entrance and exit are ordinary traces like "-> rob:Example.hs:3:1-3:15" and "<-
+bob:Example.hs:1:1-1:13" with "->" and "<-" prefixies, where "bob" and "rob" is the name
+of the function with source span. If regular script with no entrance/exit marker is given, this
+emitter will behave identically to 'logEmitter'.
+
+When script evaluation fails, this emitter will give call trace of the functions that led to the
+evaluation failure.  This is useful for pin-pointing specific area of the codebase that caused
+failure when debugging a script. When script evaluation passes, every trace message generated by
+`profile-all` flag will be removed, and this emitter will behave identically to 'logEmitter'.
+-}
+logWithCallTraceEmitter :: EmitterMode uni fun
+logWithCallTraceEmitter = EmitterMode $ \_ -> do
+  logsRef <- newSTRef DList.empty
+  let
+    addTrace DList.Nil logs = logs
+    addTrace newLogs DList.Nil = newLogs
+    addTrace newLogs logs = DList.fromList $ go (DList.toList newLogs) (DList.toList logs)
+      where
+        go l [] = l
+        go [] l = l
+        go (x:xs) l =
+          case (T.words (last l), T.words x) of
+            ("->":enterRest, "<-":exitRest) | enterRest == exitRest -> go xs (init l)
+            _                                                       -> go xs (l <> [x])
```

Comment:
Fourmolu changes this code to:
![image](https://github.com/user-attachments/assets/465fdd05-c593-44ce-9d8b-3d48a75a73cb)